src/backends/plonky2/primitives/merkletree/mod.rs

```diff
@@ -82,30 +76,83 @@ impl MerkleTree {
         &mut self,
         key: &RawValue,
         value: &RawValue,
-    ) -> TreeResult<MerkleProofStateTransition> {
+    ) -> TreeResult<MerkleTreeStateTransitionProof> {
         let proof_non_existence = self.prove_nonexistence(key)?;
 
         let old_root: Hash = self.root.hash();
         self.root
-            .add_leaf(0, self.max_depth, Leaf::new(self.max_depth, *key, *value)?)?;
+            .apply_op(self.max_depth, MerkleTreeOp::Insert, *key, Some(*value))?;
         let new_root = self.root.compute_hash();
 
         let (v, proof) = self.prove(key)?;
         assert!(proof.existence);
         assert_eq!(v, *value);
         assert!(proof.other_leaf.is_none());
 
-        Ok(MerkleProofStateTransition {
-            typ: 0, // insertion
+        Ok(MerkleTreeStateTransitionProof {
+            op: MerkleTreeOp::Insert, // insertion
             old_root,
-            proof_non_existence,
+            op_proof: proof_non_existence,
             new_root,
-            new_key: *key,
-            new_value: *value,
+            op_key: *key,
+            op_value: *value,
+            value: None,
             siblings: proof.siblings,
         })
     }
 
+    pub fn update(
+        &mut self,
+        key: &RawValue,
+        value: &RawValue,
+    ) -> TreeResult<MerkleTreeStateTransitionProof> {
+        let (old_value, old_proof) = self.prove(key)?;
+
+        let old_root: Hash = self.root.hash();
+        self.root
+            .apply_op(self.max_depth, MerkleTreeOp::Update, *key, Some(*value))?;
+        let new_root = self.root.compute_hash();
+
+        let (v, proof) = self.prove(key)?;
+        assert!(proof.existence);
+        assert_eq!(v, *value);
+        assert!(proof.other_leaf.is_none());
```

Comment:
Does `apply_op` with `Update` guarantee that we the entry at `key` exists?  which also implies that this `proof.other_leaf.is_none() = true`?
Also, `proof.siblings == old_proof.siblings` right?

I see that all the operations follow this pattern of making another proof after the `apply_op`.  This is not a suggestion for any change, I just want to confirm my understanding of the behavior.